New page creation UX

apedroferreira opened the pull request:
- File explorer-like creation UX for new pages, similar to what we have / will have for functions.
- Ability to rename pages and page hierarchy elements by double-clicking, using the same logic/component (will also have this later in functions)
- Sticky headers for these left-side panels, as opposed to the previous collapsible headers, and adjusted spacing/sizing

https://github.com/mui/mui-toolpad/assets/10789765/1e301b55-e1f0-4dfd-a288-0d504ba0139a

apedroferreira commented:
Please don't review yet, need to change the page creation tests.

Janpot commented:
cc @flaviendelangle might be interested in this use case, for an x-tree-view demo or for pro features.
* create/edit tree items
* scroll into view after creation

Janpot commented on packages/toolpad-app/src/toolpad/AppEditor/PagesExplorer/index.tsx:
```diff
@@ -166,22 +193,73 @@ export default function PagesExplorer({ className }: PagesExplorerProps) {
+
+    handleCloseCreateNewPage();
+
+    setTimeout(() => {
```
`setTimeout` doesn't feel 100% right to me. I mean, I don't think there is a strong guarantee that the item has rendered. If it needs to run after a render, shouldn't it need to run in an effect?

apedroferreira replied:
This usually always works in my experience as I just want it to wait for the next render. I did it with effects but in both cases I needed to use state just for this.

Janpot commented on packages/toolpad-app/src/toolpad/AppEditor/PagesExplorer/index.tsx:
```diff
@@ -166,22 +193,73 @@ export default function PagesExplorer({ className }: PagesExplorerProps) {
+
+  const [anchorEl, setAnchorEl] = React.useState<HTMLButtonElement | null>(null);
+  const createNewInputRef = React.useRef(null);
+  const open = !!anchorEl;
```
`anchorEl` is being reset in an affect when the `newPageInputErrorMsg` turns `null`. This means there is 1 render where `newPageInputErrorMsg` is `null` and `open` is `true`. Could that be the cause of the glitch we see in the video where there is just the icon in the popover?

apedroferreira replied:
Yeah this code was copied from the functions explorer so I saw that bug but hadn't really fixed it, should be good now.

Janpot commented on packages/toolpad-app/src/toolpad/AppEditor/PagesExplorer/index.tsx:
```diff
@@ -250,8 +329,65 @@ export default function PagesExplorer({ className }: PagesExplorerProps) {
-          onCreate={handleCreatePageDialogOpen}
+          onCreate={handleCreateNewPage}
         >
+          {isCreateNewPageOpen ? (
```
Between this and https://github.com/mui/mui-toolpad/pull/2690 is there any opportunity to share logic? I feel like in several places we need an elaborate tree component with create/edit support

bharatkashyap replied:
Agreed on the need to have a common tree component for our "explorer" use cases with the editor. It'll take some planning and dedicated effort to bridge the gap, since currently even the components for the Hierarchy explorer and the (proposed) Queries explorer are separate. Does it make sense to create an issue to track merging all tree components into one?

apedroferreira replied:
I've created a `CreateTreeItem` component for now - should be able to use it in the global functions explorer too.

apedroferreira replied:
> Does it make sense to create an issue to track merging all tree components into one?

If we want to create a whole tree view component with our own setting and features to reuse everywhere we can create a separate issue for it and do it, feels a bit early for this PR but should be a good idea long term.

bharatkashyap replied:
Adding it as a discussion item for next to next grooming

Janpot commented on packages/toolpad-app/src/components/CreateTreeItem.tsx:
```diff
@@ -0,0 +1,151 @@
+  treeRef,
+  open,
+  suggestedNewItemName = '',
+  onCreate,
```
Not sure how far to take this here. But this component is very focused on the "create" use case, while I'd expect the create logic to share 90% with the edit logic. Would an `EditableTreeItem` which is agnostic of whether it's in a create or edit scenario not make more sense?

apedroferreira replied:
Should I try to make page names editable in the pages explorer maybe, to see how much of the logic I can share and how? And remove the ability to do it from the right-side panel? I guess that most of this logic might be sharable, but there is probably some more specific logic that I'd have to see how it fits.

Janpot commented:
> What's the reason for using a TreeView in the first place given that pages can't be nested, and collapsing the top-level doesn't seem useful?

We're looking for a UX similar to the vscode file explorer. The data roughly maps on the file system and has similar functionality around create/rename. It's true that pages can't be nested (yet), we do plan to use this pattern in some places in our UI which will have nested items. So far the other places have 1 level of nesting (function files with exported functions). It could be valuable to explore whether there are other UI patterns than a tree view to represent these.

We should probably remove the top-level "Pages" item. This is a leftover from a time when there were more top-level items.

Janpot commented on packages/toolpad-app/src/toolpad/AppEditor/PagesExplorer/index.tsx:
```diff
@@ -224,41 +337,49 @@ export default function PagesExplorer({ className }: PagesExplorerProps) {
+            sx={pagesTreeItemSx}
+          />
+        ) : null}
+        {sortBy(pages, 'name').map((page) => (
```
This is mutating the `pages` array. Best to Avoid mutating variables in the render function.
It's also running a sort i every render. Feels a bit wasteful.

apedroferreira replied:
Yeah, I noticed it was mutable, in this particular case it doesn't have too much impact but it's better to make it immutable and memoize it, I'll do it.

Janpot commented on packages/toolpad-utils/src/collections.ts:
```diff
@@ -108,3 +108,18 @@ export function equalProperties<P extends object>(obj1: P, obj2: P, subset?: (ke
+ * Sorts an array of objects by the value of one property.
+ */
+export function sortBy<P extends object>(objs: P[], prop: keyof P, isAsc = true): P[] {
+  return objs.sort((a, b) => {
```
None of the other exports of this file mutate the source array. Feels a bit counterintuitive, and I feel like it could lead to bugs. An alternative could be to add a 

```tsx
export function createPropertyComparator<T>(propName: string): (a: T, b: T) => number;
```

and let the consumer decide whether they want to mutate their array or not.

apedroferreira replied:
I think it's ok to just make this method immutable, simpler and more intuitive. At first I guess I wanted it to be more like `.sort`, but the way that we use it it's probably better for it just to be immutable.

Janpot reviewed (commented):
Looks good. I proposed a few tweaks regarding project organization. Other remarks:

* 🤔 For visual consistency, I feel like it would be better if these are aligned

   <img width="302" alt="Screenshot 2023-10-05 at 11 11 46" src="https://github.com/mui/mui-toolpad/assets/2109932/fb4e5f58-9761-4414-8625-0f5d9caffd86">

* Will it be confusing to remove the existing page name field? It's a bit obsolete now, but I feel like we could still offer a way to edit the name in that page options panel. Perhaps less obtrusive as the one we have now. In any case, we can keep it for a follow up PR if we leave the name existing editor in place.

apedroferreira commented:
> * 🤔 For visual consistency, I feel like it would be better if these are aligned

The issue is that in the hierarchy explorer we need space for the little expand/collapse arrow for nestable elements. I'll see if it works to add some more padding to the pages though, might be fine.

> * Will it be confusing to remove the existing page name field? It's a bit obsolete now, but I feel like we could still offer a way to edit the name in that page options panel. Perhaps less obtrusive as the one we have now. In any case, we can keep it for a follow up PR if we leave the name existing editor in place.

I've actually already removed it... There's pros and cons to removing it or not, on one hand it gives the users maybe a more obvious place to change the name, but on the other hand it might be confusing how it's the same thing as renaming on the left. And it might be easier for users to associate renaming in the left panel with renaming the project files. I guess I can keep the input for now and we'll see how things go from there.

apedroferreira commented on packages/toolpad-app/src/utils/scroll.ts:
That's fine, I wasn't really sure where to put this.

apedroferreira commented on packages/toolpad-app/src/components/ExplorerHeader.tsx:
No, I'm using in both the hierarchy explorer and the pages explorer. It's supposed to be the header for any "explorer" panel like this.
Eventually I can make these components part of a larger one that handles any explorer panel including the header and creatable/renamable items. I guess I wanted too keep it more modular first. But maybe I should try it already.

apedroferreira replied:
Oh, I can try to move the component above just enough in the project structure probably if it makes sense, might be better than in the `components` folder then.

apedroferreira commented:
I've added all the changes discussed. Updated UI screenshot:

<img width="1792" alt="Screenshot 2023-10-05 at 20 44 46" src="https://github.com/mui/mui-toolpad/assets/10789765/a1ea3ef9-aed7-4df7-ba4f-0a5232103869">

Janpot commented on packages/toolpad-app/src/toolpad/AppEditor/PagesExplorer/index.tsx:
```diff
@@ -321,11 +321,7 @@ export default function PagesExplorer({ className }: PagesExplorerProps) {
-      ml: -1.5,
-    },
-  };
+  const alphabeticallySortedPages = React.useMemo(() => sortBy(pages, 'name'), [pages]);
```
That's not alphabetically sorted. e.g. `'a' < 'B' === false`.

Not a fan of `sortBy`, I'd just do:

```tsx
[..pages].sort((page1, page2) => page1.name.localeCompare(page2.name)
```

apedroferreira replied:
Uh, that's right about the alphabetical ordering... And nice, I didn't know about `localeCompare` or `toSorted`.
The first approach seems fine, I tried implementing the second abstraction you proposed but I'm not really sure how to? And we might not need that abstraction yet anyway...

Janpot reviewed (approved):
Great, thanks for updating!

Janpot commented:
@apedroferreira The position of the scrollbar seems to be off on the pages section. I don't believe position: sticky/fixed should be used for headers. The UX feels off when the scrollbar extends over the header.

<img width="304" alt="Screenshot 2023-10-07 at 01 36 54" src="https://github.com/mui/mui-toolpad/assets/2109932/900caa4f-14c6-4191-a044-68c3e28c9946">
